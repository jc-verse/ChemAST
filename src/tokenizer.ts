function split(formula: string): string[] {
  const segments = [];
  for (let i = 0; i < formula.length; ) {
    if (["*", "(", "[", "{", ")", "]", "}"].includes(formula[i])) {
      segments.push(formula[i]);
      i++;
    } else if (/[A-Z]/.test(formula[i])) {
      if (/[a-z]/.test(formula[i + 1])) {
        segments.push(formula.substring(i, i + 2));
        i += 2;
      } else {
        segments.push(formula.substring(i, i + 1));
        i++;
      }
    } else if (/[0-9]/.test(formula[i])) {
      let num = formula[i];
      i++;
      while (/[0-9]/.test(formula[i])) {
        num += formula[i];
        i++;
      }
      segments.push(num);
    } else if (formula[i] === " ") {
      i++;
    } else {
      throw new Error(`Parsing Error at position ${i}: '${formula[i]}'`);
    }
  }
  return segments;
}

export function tokenize(formula: string): Token[] {
  const segments = split(formula);
  return segments.map((seg) => {
    if (seg === "(" || seg === "[" || seg === "{") {
      return { type: "lParen", value: seg };
    } else if (seg === ")" || seg === "]" || seg === "}") {
      return { type: "rParen", value: seg };
    } else if (seg === "*") {
      return { type: "join", value: seg };
    } else if (/\d+/.test(seg)) {
      return { type: "number", value: Number(seg) };
    } else {
      return { type: "atom", value: seg };
    }
  });
}
