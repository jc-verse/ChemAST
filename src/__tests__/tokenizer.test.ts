import {tokenize} from '../tokenizer';

describe('tokenize', () => {
  test('Should accept valid formulae', () => {
    expect(tokenize('CuSO4 * 5H2O')).toMatchSnapshot();
    expect(tokenize('(NH4)3PO4')).toMatchSnapshot();
    expect(tokenize('CaCO3')).toMatchSnapshot();
    expect(tokenize('[Cu(NH3)4]SO4')).toMatchSnapshot();
    expect(tokenize('CO')).toMatchSnapshot();
    expect(tokenize('2Cu')).toMatchSnapshot();
    expect(tokenize('CaCaCa')).toMatchSnapshot();
    expect(tokenize('CCC')).toMatchSnapshot();
    expect(tokenize('5H2O*(H2O2)2')).toMatchSnapshot();
  });

  test('Should reject bad formulae', () => {
    expect(() => tokenize('Sooo')).toThrowErrorMatchingInlineSnapshot(`"Parsing Error at position 2: 'o'"`);
  })
})
