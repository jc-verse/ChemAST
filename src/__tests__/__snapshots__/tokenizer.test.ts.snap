// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenize Should accept valid formulae 1`] = `
Array [
  Object {
    "type": "atom",
    "value": "Cu",
  },
  Object {
    "type": "atom",
    "value": "S",
  },
  Object {
    "type": "atom",
    "value": "O",
  },
  Object {
    "type": "number",
    "value": 4,
  },
  Object {
    "type": "join",
    "value": "*",
  },
  Object {
    "type": "number",
    "value": 5,
  },
  Object {
    "type": "atom",
    "value": "H",
  },
  Object {
    "type": "number",
    "value": 2,
  },
  Object {
    "type": "atom",
    "value": "O",
  },
]
`;

exports[`tokenize Should accept valid formulae 2`] = `
Array [
  Object {
    "type": "lParen",
    "value": "(",
  },
  Object {
    "type": "atom",
    "value": "N",
  },
  Object {
    "type": "atom",
    "value": "H",
  },
  Object {
    "type": "number",
    "value": 4,
  },
  Object {
    "type": "rParen",
    "value": ")",
  },
  Object {
    "type": "number",
    "value": 3,
  },
  Object {
    "type": "atom",
    "value": "P",
  },
  Object {
    "type": "atom",
    "value": "O",
  },
  Object {
    "type": "number",
    "value": 4,
  },
]
`;

exports[`tokenize Should accept valid formulae 3`] = `
Array [
  Object {
    "type": "atom",
    "value": "Ca",
  },
  Object {
    "type": "atom",
    "value": "C",
  },
  Object {
    "type": "atom",
    "value": "O",
  },
  Object {
    "type": "number",
    "value": 3,
  },
]
`;

exports[`tokenize Should accept valid formulae 4`] = `
Array [
  Object {
    "type": "lParen",
    "value": "[",
  },
  Object {
    "type": "atom",
    "value": "Cu",
  },
  Object {
    "type": "lParen",
    "value": "(",
  },
  Object {
    "type": "atom",
    "value": "N",
  },
  Object {
    "type": "atom",
    "value": "H",
  },
  Object {
    "type": "number",
    "value": 3,
  },
  Object {
    "type": "rParen",
    "value": ")",
  },
  Object {
    "type": "number",
    "value": 4,
  },
  Object {
    "type": "rParen",
    "value": "]",
  },
  Object {
    "type": "atom",
    "value": "S",
  },
  Object {
    "type": "atom",
    "value": "O",
  },
  Object {
    "type": "number",
    "value": 4,
  },
]
`;

exports[`tokenize Should accept valid formulae 5`] = `
Array [
  Object {
    "type": "atom",
    "value": "C",
  },
  Object {
    "type": "atom",
    "value": "O",
  },
]
`;

exports[`tokenize Should accept valid formulae 6`] = `
Array [
  Object {
    "type": "number",
    "value": 2,
  },
  Object {
    "type": "atom",
    "value": "Cu",
  },
]
`;

exports[`tokenize Should accept valid formulae 7`] = `
Array [
  Object {
    "type": "atom",
    "value": "Ca",
  },
  Object {
    "type": "atom",
    "value": "Ca",
  },
  Object {
    "type": "atom",
    "value": "Ca",
  },
]
`;

exports[`tokenize Should accept valid formulae 8`] = `
Array [
  Object {
    "type": "atom",
    "value": "C",
  },
  Object {
    "type": "atom",
    "value": "C",
  },
  Object {
    "type": "atom",
    "value": "C",
  },
]
`;

exports[`tokenize Should accept valid formulae 9`] = `
Array [
  Object {
    "type": "number",
    "value": 5,
  },
  Object {
    "type": "atom",
    "value": "H",
  },
  Object {
    "type": "number",
    "value": 2,
  },
  Object {
    "type": "atom",
    "value": "O",
  },
  Object {
    "type": "join",
    "value": "*",
  },
  Object {
    "type": "lParen",
    "value": "(",
  },
  Object {
    "type": "atom",
    "value": "H",
  },
  Object {
    "type": "number",
    "value": 2,
  },
  Object {
    "type": "atom",
    "value": "O",
  },
  Object {
    "type": "number",
    "value": 2,
  },
  Object {
    "type": "rParen",
    "value": ")",
  },
  Object {
    "type": "number",
    "value": 2,
  },
]
`;
